modules/dlsobackend/dlsobackend.cc

```diff
@@ -0,0 +1,722 @@
+    break;
+  default:
+  case DOMAIN_INFO_KIND_MASTER:
+    di->kind = DomainInfo::Master;
+    break;
+  }
+}
+
+bool DlsoBackend::getDomainInfo(const DNSName& domain, DomainInfo& di, bool getSerial)
+{
+  if (api->get_domain_info == nullptr)
+    return false;
+
+  string qname = domain.toString();
+  di.backend = this;
+
+  return api->get_domain_info(api->handle, qname.size(), qname.c_str(), fill_domain_info, &di);
+}
+
+bool DlsoBackend::startTransaction(const DNSName& domain, int domain_id)
+{
+  if (api->start_transaction == nullptr)
+    return false;
+
+  string domain_ = domain.toString();
+
+  return api->start_transaction(api->handle, domain_id, domain_.size(), domain_.c_str());
+}
+
+bool DlsoBackend::abortTransaction()
+{
+  if (api->abort_transaction == nullptr)
+    return false;
+  return api->abort_transaction(api->handle);
+}
+
+bool DlsoBackend::commitTransaction()
+{
+  if (api->commit_transaction == nullptr)
+    return false;
+  return api->commit_transaction(api->handle);
+}
+
+void fill_unfresh_slave(const void* unfresh_, struct domain_info* domain_info)
+{
+  vector<DomainInfo>* unfresh = (vector<DomainInfo>*)unfresh_;
+
+  DomainInfo di;
+  fill_domain_info(&di, domain_info);
+
+  unfresh->push_back(di);
+}
+
+void DlsoBackend::getUnfreshSlaveInfos(vector<DomainInfo>* unfreshDomains)
+{
+  if (api->get_unfresh_slave == nullptr)
+    return;
+
+  bool status = api->get_unfresh_slave(api->handle, fill_unfresh_slave, unfreshDomains);
+
+  if (!status) {
+    throw PDNSException("DlsoBackend unable to retrieve list of slave domains");
+  }
+}
+
+void DlsoBackend::setNotified(uint32_t domain_id, uint32_t serial)
+{
+  if (api->set_notified == nullptr)
+    return;
+
+  api->set_notified(api->handle, domain_id, serial);
+}
+
+void DlsoBackend::setFresh(uint32_t domain_id)
+{
+  if (api->set_fresh == nullptr)
+    return;
+
+  api->set_fresh(api->handle, domain_id);
+}
+
+bool DlsoBackend::replaceRRSet(uint32_t domain_id, const DNSName& qname, const QType& qt, const vector<DNSResourceRecord>& rrset)
+{
+  if (api->replace_record == nullptr)
+    return false;
+
+  vector<string> qnames;
+  struct resource_record* records = (struct resource_record*)calloc(rrset.size(), sizeof(struct resource_record));
+  uint32_t i = 0;
+
+  for (const auto& rr : rrset) {
+    qnames.push_back(rr.qname.toString());
+
+    const string qname_ = qnames.back();
+
+    records[i].qtype = rr.qtype.getCode();
+    records[i].qname = qname_.c_str();
+    records[i].qname_len = qname_.size();
+    records[i].content = rr.content.c_str();
+    records[i].content_len = rr.content.size();
+    records[i].ttl = rr.ttl;
+    records[i].auth = rr.auth;
+    records[i].scope_mask = rr.scopeMask;
+    records[i].domain_id = rr.domain_id;
+
+    i++;
+  }
+
+  string qname_ = qname.toString();
+
+  try {
+    bool status = api->replace_record(api->handle, domain_id, qname_.size(), qname_.c_str(), qt.getCode(), rrset.size(), records);
```

Comment:
I suppose using a `vector` ctor+reserve, and give access to the 3rd party lib to ptr with `vector::data()` + size https://en.cppreference.com/w/cpp/container/vector/data may be fine?

Reply:
Using a `vector`, either passing the final size to the constructor or using `resize()` (`reserve()` does not guarantee anything), would be the right way to go, yes.